Objects/object.c

```diff
@@ -230,6 +230,9 @@ PyObject_Init(PyObject *op, PyTypeObject *tp)
         return PyErr_NoMemory();
     /* Any changes should be reflected in PyObject_INIT (objimpl.h) */
     Py_TYPE(op) = tp;
+    if (PyType_GetFlags(tp) & Py_TPFLAGS_HEAPTYPE) {
```

Comment:
I am not sure about this part. It will affect all other heap types. It is better to defer it to a separate issue.

Reply:
This is really needed for this change. PyTypeObjects should participate in reference counting. Without this change, instantiating a `DirEntry` or a `ScandirIterator` will not increase the type's refcount - which is incorrect.

This ensures the correct behavior to the newly heap-allocated type. I added it in this place to ensure that any future migration from PyType_Ready to PyType_FromSpec won't have to worry about this anymore.

Reply:
To expand on this a bit more, this currently **DOES NOT** affect any other heap types but the ones changed here (with 1 exception).

There are only 3 ways to access `PyObject_{Init,INIT}` which are:
* Through `PyType_GenericAlloc`
* Through `PyObject_New` (corollary: `PyObject_NewVar`)
* Direct calls which don't come from the previous two. (corollary: Direct calls to `PyObject_Init`, `PyObject_INIT_VAR`)

For each of these points:
* All current calls to `PyType_GenericAlloc` are already checking if type type is a heap type. This just moves the check from `PyType_GenericAlloc`, to `PyObject_Init` so it preserves behavior.
* All current calls to `PyObject_New` and `PyObject_NewVar` use static types. With 1 exception, mentioned below.
* All direct calls use static types. 

Outside of the exception there should not be any problems with any other heap type.

Exception:
_tkinter.c. The incref was manually added after `PyObject_New`.

I have the following 3 proposals:
1) Leave the current change. The _tkinter types will temporary leak a refcount per instance allocation. Not that big of a deal since it's just 3 different types. Then, add a new PR to fix _tkinter.c.
2) Modify _tkinter.c in this change to do the correct thing.
3) Remove the `Py_INCREF` in `PyObject_Init` in this change and add it right after `PyObject_New `in posixmodule.c. Then create a new PR with this change to `Py_INCREF` in `PyObject_Init` and fix both posixmodule.c and _tkinter.c